stl/inc/iomanip

```diff
@@ -213,34 +213,28 @@ struct _Quote_out { // store pointer/length for string
 template <class _Elem>
 _NODISCARD _Quote_out<_Elem, void, size_t> quoted(
     const _Elem* _Ptr, _Elem _Delim = _Elem('"'), _Elem _Escape = _Elem('\\')) {
-    size_t _Size = 0;
-    while (_Ptr[_Size] != _Elem(0)) {
-        ++_Size;
-    }
-
-    return _Quote_out<_Elem, void, size_t>(_Ptr, _Size, _Delim, _Escape);
+    const size_t _Size = char_traits<_Elem>::length(_Ptr);

```

Comment:
I don't know enough about the standard to say, but if `charT` is allowed to be a non-builtin character type and someone implements `char_traits<charT>` in a stupid way, this is a change in behavior; also, if `charT`s are allowed to have stupid constructors from `int`, (or are allowed to not be constructible from `int`), then the old code was also wrong - it should use `!= _Elem()`

Reply:
More explicitly:

First, change the old code to be `_Elem()`, not `_Elem(0)` - that was a bug anyways.

Then:
let `x: _Elem` such that `char_traits<_Elem>::eq(x, _Elem())`, but `x != _Elem()`.
set `p` to be the address of an array `a = {x, _Elem()}`.
Then, in the old code (w/ bugfix), `_Size = 1`
In the new code, `_Size = 0`.
Clearly, a change in behavior has occured.

Reply:
I think it is safe to use `char_traits<_Elem>::length`.

The `_Quoted_out` returned from this function can only be used with `basic_ostream<_Elem>`, and AFAIK `basic_ostream` only supports `char` and `wchar_t` (and maybe `unsigned short`). In either case `char_traits<_Elem>` is a known specialization that does not do anything special.

Reply:
I see nothing in the standard that says `basic_ostream` can't be used with types other than `char` and `wchar_t`; it does, for example, work with char32_t in MSVC's implementation:

```cxx
#include <sstream>
#include <iostream>
using namespace std;

int main() {
	basic_stringstream<char32_t> s;
	s << U"Hello, world!\n";
	for (char32_t ch : s.str()) {
		// this is incorrect but _shrug_
		wcout << static_cast<wchar_t>(ch);
	}
}
```

and also, it works for my own char types:

```cxx
include <sstream>
#include <iostream>
#include <compare>
using namespace std;

enum class MyWcharT : wchar_t {};

constexpr bool operator==(MyWcharT lhs, MyWcharT rhs) noexcept {
        return static_cast<wchar_t>(lhs) == static_cast<wchar_t>(rhs);
}
constexpr strong_ordering operator<=>(MyWcharT lhs, MyWcharT rhs) noexcept {
        return static_cast<wchar_t>(lhs) <=> static_cast<wchar_t>(rhs);
}

template <>
class char_traits<MyWcharT> {
public:
        using char_type = MyWcharT;
        using int_type = int;
        using off_type = ptrdiff_t;
        using pos_type = ptrdiff_t;
        using state_type = std::mbstate_t;

        static constexpr void assign(char_type& r, const char_type& a) noexcept {
                r = a;
        }
        static constexpr char_type* assign(char_type* p, size_t count, char_type a) {
                for (size_t i = 0; i < count; ++i) {
                        p[i] = a;
                }
                return p;
        }
        static constexpr bool eq(char_type lhs, char_type rhs) noexcept {
                return lhs == rhs;
        }
        static constexpr bool lt(char_type lhs, char_type rhs) noexcept {
                return lhs < rhs;
        }
        static char_type* move(char_type* dest, const char_type* src, std::size_t count) {
                memmove(dest, src, count * (sizeof *dest));
                return dest;
        }
        static char_type* copy(char_type* dest, const char_type* src, std::size_t count) {
                memcpy(dest, src, count * (sizeof *dest));
                return dest;
        }
        static constexpr int compare(const char_type* s1, const char_type* s2, std::size_t count) {
                auto res = lexicographical_compare_three_way(s1, s1 + count, s2, s2 + count);
                return res < 0 ? -1 : (res > 0 ? 1 : 0);
        }
        static constexpr size_t length(const char_type* s) {
                size_t sz = 0;
                while (*s++ != MyWcharT()) ++sz;
                return sz;
        }
        static constexpr const char_type* find(const char_type* p, std::size_t count, const char_type& ch) {
                auto res = std::find(p, p + count, ch);
                if (res == p + count) return nullptr;
                return res;
        }
        static constexpr char_type to_char_type(int_type c) noexcept {
                return static_cast<MyWcharT>(c);
        }
        static constexpr int_type to_int_type(char_type c) noexcept {
                return static_cast<int>(c);
        }
        static constexpr bool eq_int_type( int_type c1, int_type c2 ) noexcept {
                return c1 == c2;
        }
        static constexpr int_type eof() noexcept {
                return -1;
        }
        static constexpr int_type not_eof(int_type e) noexcept {
                return e != -1 ? e : 0;
        }
};

int main() {
        basic_stringstream<MyWcharT> s;
        const MyWcharT str[] = {
                static_cast<MyWcharT>(L'H'),
                static_cast<MyWcharT>(L'i'),
                static_cast<MyWcharT>(L'!'),
                static_cast<MyWcharT>(L'\n'),
                static_cast<MyWcharT>(L'\0'),
        };

        s << str;

        for (MyWcharT ch : s.str()) {
                wcout << static_cast<wchar_t>(ch);
        }
}
```

Reply:
I'm okay with this resolution, though I'd like to open an issue.

Reply:
I have filed #2691 as a reminder to file an LWG issue. Thanks!